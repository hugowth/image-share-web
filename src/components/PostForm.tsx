import React from "react";
import { Formik, Form, Field } from "formik";
import { Button, Box, Grid, TextField } from "@material-ui/core";
import { PostPayload } from "../types";
import { object, string, mixed } from "yup";

type Props = {
  onSubmit: (values: PostPayload) => Promise<void>;
};

const FILE_SIZE = 160 * 1024;
const SUPPORTED_FORMATS = ["image/jpg", "image/jpeg", "image/gif", "image/png"];

const schema = object({
  image: mixed()
    .required("A file is required")
    .test(
      "fileSize",
      "File too large",
      (value) => value && value.size <= FILE_SIZE
    )
    .test(
      "fileFormat",
      "Unsupported Format",
      (value) => value && SUPPORTED_FORMATS.includes(value.type)
    ),
  description: string().required(),
});

const PostForm = ({ onSubmit }: Props) => {
  const fileInput = React.createRef();
  return (
    <>
      <Formik
        initialValues={{ image: undefined, description: "" }}
        onSubmit={onSubmit}
        validationSchema={schema}
        validateOnMount
      >
        {(formik) => {
          const { isSubmitting, isValid, errors, touched, setFieldValue } =
            formik;
          const disabledField = isSubmitting;
          const disabledSubmmit = !isValid || isSubmitting;

          return (
            <Form>
              <Grid container spacing={3}>
                <Grid item xs={12}>
                  <input
                    type="file"
                    style={{ display: "none" }}
                    onChange={(event) => {
                      // @ts-ignore
                      setFieldValue("image", event.currentTarget.files[0]);
                    }}
                    // @ts-ignore
                    ref={fileInput}
                  />

                  <button
                    type="button"
                    // @ts-ignore
                    onClick={() => fileInput.current.click()}
                  >
                    Choose file
                  </button>
                  <>{errors.image && errors.image}</>
                </Grid>

                <Grid item xs={12}>
                  <Field
                    as={TextField}
                    variant="outlined"
                    fullWidth
                    autoFocus
                    name="description"
                    label="description"
                    error={touched.description && Boolean(errors.description)}
                    helperText={errors.description}
                    disabled={disabledField}
                  />
                </Grid>
              </Grid>

              <Box my={2}>
                <Button
                  type="submit"
                  variant="outlined"
                  color="primary"
                  disabled={disabledSubmmit}
                >
                  Submit
                </Button>
              </Box>
            </Form>
          );
        }}
      </Formik>
    </>
  );
};

export default PostForm;
