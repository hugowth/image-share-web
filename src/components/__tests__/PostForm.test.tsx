import React from "react";
import renderer from "react-test-renderer";

import PostForm from "../PostForm";

describe("PostForm", () => {
  it("renders correctly", async () => {
    const onSubmit = jest.fn();
    const tree = renderer.create(<PostForm onSubmit={onSubmit} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
