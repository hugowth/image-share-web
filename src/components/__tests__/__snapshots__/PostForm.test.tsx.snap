// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm renders correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <input
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
      <button
        onClick={[Function]}
        type="button"
      >
        Choose file
      </button>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
        >
          description
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            name="description"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                description
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiBox-root MuiBox-root-5"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Submit
      </span>
    </button>
  </div>
</form>
`;
